File: lc33.cc
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int lo = 0, hi = nums.size() - 1;
        int mid;
        while(lo < hi) {
    		mid = (lo + hi) / 2;
    		if(target == nums[mid]) {
    			return mid;
    		}
        	if(nums[lo] < nums[hi]) {
        		if(target < nums[mid]) {
        			hi = mid - 1;
        		} else {
        			lo = mid + 1;
        		}
        	} else {
        		if(target > nums[mid]) {
        			if(nums[mid] > nums[hi]) {
        				lo = mid + 1;
        			} else {
        				if(target <= nums[hi]) {
	        				lo = mid + 1;
        				} else {
	        				hi = mid - 1;
        				}
        			}
        		} else {
        			//target < nums[mid]
        			if(nums[mid] < nums[lo]) {
        				hi = mid - 1;
        			} else {
        				if(target >= nums[lo]) {
	        				hi = mid - 1;
        				} else {
	        				lo = mid + 1;
        				}
        			}
        		}
        	}
        }
        if(nums[lo] == target) return lo;
        return -1;
    }
};

File: lc33.java
public class Solution {
    public int search(int[] nums, int target) {    
        int l = 0, r = nums.length - 1;
        while(l < r) {
            int m = l + (r - l) / 2;
            if(target == nums[m]) return m;
            if(target > nums[m]) {
                if(nums[l] < nums[r]) {
                    //in order
                    l = m + 1;
                }
                else if(nums[r] >= target) {
                    //in right part
                    l = m + 1;
                }
                else { //if(nums[r] < target) {
                    // may be in the left
                    if(nums[l] < nums[m]) {
                        //max in the right 
                        l = m + 1;
                    }
                    else 
                        r = m - 1;
                }
            } else { // target < nums[m]
                if(nums[l] < nums[r]) {
                    r = m - 1;
                }
                else if(nums[l] > target) {
                    if(nums[r] > nums[m])
                        r = m - 1;
                    else
                        l = m + 1;
                } else {
                    r = m - 1;
                }
            }
        }
        if(nums[l] == target) return l;
        else return -1;
    }    
}